src/coreclr/jit/utils.cpp

```diff
@@ -2272,8 +2272,13 @@ double FloatingPointUtils::convertUInt64ToDouble(uint64_t uIntVal)
 
 float FloatingPointUtils::convertUInt64ToFloat(uint64_t u64)
 {
+#ifdef TARGET_ARM64
+    // ARM64 supports casting directly to float
+    return (float)u64;
+#else  // !TARGET_ARM64
     double d = convertUInt64ToDouble(u64);
     return (float)d;
+#endif // !TARGET_ARM64
```

Comment:
I'm pretty sure this bug exists on all platforms, not just Arm64.

The code in the test is
```
ulong vr10 = 16105307123914158031UL;
float vr11 = 4294967295U | vr10;
```

This boils down to `4294967295U | 16105307123914158031UL` which is `16105307124325679103`

The correct `float` result is then `16105306574569865216.0f`, or rather the raw bits `1600094603` (the DEBUG output).

The result produced by the two step conversion, however, is `16105307674081492992.0f`, or rather the raw bits `1600094604` (the RELEASE output).

I would expect that all our current compilers (MSVC, Clang, and GCC) are producing correct results for `return (float)u64` and we no longer need to do a two step conversion. If any are still producing incorrect results, we should log bugs against them and we'd need to hand roll an implementation that does the conversion instead (I can give reference to one if we need it, but I don't think we do).

Reply:
Thanks for pointing this out. I tried both the one- and two-step conversion on x64, and in Debug and Release, I get `1600094604`. So it looks like the MSVC/Clang/GCC codegen on the const-folding path is matching what RyuJIT currently emits in Debug for x64. For this test case, is the definition of "correct" architecture-dependent? I'm guessing there's some subtle difference in rounding behavior between ARM64 and x64 that's leading to this diff in behavior?

Regardless of the answer to that, I think you're right that we can update the cast logic for all platforms. I can do that in this PR.

Reply:
> I tried both the one- and two-step conversion on x64, and in Debug and Release, I get 1600094604

Are you sure? I don't see the same behavior in godbolt: https://godbolt.org/z/s5MrTWaYM
-- note that `m1` and `m2` return the same result (raw bits `0x5f5f818b`) while `m3` returns one higher (raw bits `0x5f5f818c`)

> For this test case, is the definition of "correct" architecture-dependent?

No, we should be deterministic across all platforms for such a case; and correspondingly the added test should be enabled for all platforms as well, not just be Arm64 specific

Reply:
It turns out the `ulong -> float` cast on x64 is represented in IR as `ulong -> double -> float`, while on ARM64, we combine it into one cast during morph. I'm guessing we have to enable that path for all platforms, too.

Reply:
So the importer is creating `ulong -> double -> float` for `ulong -> float` conversions in IL? That sounds wrong given that those are not equivalent (as the example shows).

Reply:
Yeah, this is the IR after importation:
```
***** BB01 [0000]
STMT00000 ( 0x000[E-] ... 0x009 )
               [000001] DA---------                         *  STORE_LCL_VAR long   V00 loc0         
               [000000] -----------                         \--*  CNS_INT   long   -0x207e748018875031

***** BB01 [0000]
STMT00001 ( 0x00A[E-] ... 0x01A )
               [000009] --C-G------                         *  CALL      void   System.Console:WriteLine(uint)
               [000008] ----------- arg0                    \--*  BITCAST   int   
               [000007] -----------                            \--*  CAST      float <- double
               [000006] ---------U-                               \--*  CAST      double <- ulong
               [000005] -----------                                  \--*  OR        long  
               [000003] -----------                                     +--*  CNS_INT   long   0xFFFFFFFF
               [000004] -----------                                     \--*  LCL_VAR   long   V00 loc0         

***** BB01 [0000]
STMT00002 ( 0x01A[E-] ... ??? )
               [000010] -----------                         *  RETURN    void 
```

I'll pinpoint where we're making this decision...

Reply:
Hmm, it seems like it's not really possible to represent a `ulong -> float` conversion in IL. The `conv.r.un` Roslyn is using here is the only unsigned conversion available, but it is presumably what ends up resulting in the conversion to double. I'm not really sure what can be done here except consistently make the result the same as `(float)(double)value` -- this seems like a flaw in the spec.

Reply:
> I'm not really sure what can be done here except consistently make the result the same as (float)(double)value -- this seems like a flaw in the spec.

It sounds like we should morph `ulong -> double -> float` into `ulong -> float` on all platforms for now, then. Since it's early in the .NET 10 cycle, is it possible to propose fixing this in the spec?

Reply:
The `ulong -> double -> float` to `ulong -> float` transformation is not value preserving, so it doesn't seem right to me that we're doing that on internal IR. But if we're looking for a .NET 9 fix then maybe that is the most surgical.

For a fix it seems like we would need to provide some conversion helpers that Roslyn could use (or Roslyn would need to emit some more manual IL to do the conversion).

Reply:
Roslyn (C#, VB) and F# all differentiate a bit in IL

For `ulong->double` they emits `conv.r.un, conv.r8`
while for `ulong->float` they  emit `conv.r.un, conv.r4`
and for `ulong->double->float` they emit `conv.r.un, conv.r8, conv.r4`

I think this is enough that we can sufficiently recognize the intent of the IR and we can treat just `conv.r.un`  and `conv.r.un, conv.r8` as `convert to double`, but `conv.r.un, conv.r4` as `convert to float`.

If we follow that, we can do the "intended" thing on import for the vast majority of code out there without everyone needing to recompile. We would then not do any folding of `ulong->double->float` in morph since that is incorrect and does the wrong thing if the user intended to do a 2 step conversion (as the IL would have indicated)

Reply:
While I agree that it would be more practical I'm really not a fan of how that means we redefine the meaning of a particular sequence of IL instructions. That breaks composability of the language.

Reply:
I'm not sure it's really redefining it. We've historically done a lot of weird things here and treated it differently based on what the target hardware supports That is, this isn't a regression new to .NET 9, but rather a longstanding behavior. Here's where some of it was special-cased in .NET 5 (https://github.com/dotnet/runtime/blob/v5.0.0/src/coreclr/src/jit/morph.cpp#L265-L355), but it goes back well before that, I believe even into .NET Framework.

In .NET 9, however, we did make explicit efforts to fix known floating-point to/from integer conversion bugs and ensure xplat determinism. So by normalizing the behavior to match what Roslyn (C#, VB) and F# intended from their source, it's rather inline with the general push we made this release and matching what the typical developer would expect given what they authored. It might redefine some manual IL or some theoretical 3rd party language, but I think its a pretty simple thing to document and explain and doesn't require us to further bloat IL or leave the world split between two states based on whether they recompile with the latest Roslyn or F# compilers. Instead, they have something that is clear and works without needing to recompile given the long-standing set of 3 patterns the higher level compilers have emitted

Reply:
It would mean that
```
conv.r.un
conv.r4
```
would not give the same result as
```
conv.r.un
nop
conv.r4
```
The nop here can be an arbitrary sequence of instructions or even branches to other basic blocks. That's the part I dislike, and why I would prefer some helper that does the conversion explicitly.

Reply:
How likely is that code to exist in the real world however? I don’t think Roslyn or F# emits such code today, for example (as best as I can tell, they always emit conv.r.un as a pair since it isn’t strictly either double or float as a result, by spec)

A helper could certainly make it more guaranteed, but has its own cost and negatives as well.

Reply:
One clear negative, for example, is it becomes impossible to get the desired behavior while targeting netstandard2.0

You become forced to multi target for the C#/F# code you write to not have a bug on modern .NET

Reply:
I think trying to get around Roslyn's two-step cast behavior is going to be trickier than pattern-matching the IL. If I get rid of the first local in the initial example to make the bitwise-OR trivially foldable, like so:
```csharp
public class Program
{
    public static void Main()
    {
        float vr11 = 4294967295U | 16105307123914158031UL;
        System.Console.WriteLine(System.BitConverter.SingleToUInt32Bits(vr11));
    }
}
```

Roslyn does the two-step cast, and emits the following IL:
```
IL_0000  22 8c 81 5f 5f    ldc.r4       16105307674081492992.000000
IL_0005  28 0d 00 00 0a    call         0xA00000D
IL_000a  28 0e 00 00 0a    call         0xA00000E
IL_000f  2a                ret
```

And so we import the following IR:
```
STMT00000 ( 0x000[E-] ... ??? )
               [000002] --C-G------                         *  CALL      void   System.Console:WriteLine(uint)
               [000001] ----------- arg0                    \--*  CNS_INT   int    0x5F5F818C

    [ 0]  15 (0x00f) ret

STMT00001 ( 0x00F[E-] ... ??? )
               [000003] -----------                         *  RETURN    void
```

Reply:
Roslyn is likely relying on the JIT implementation here, not realizing it's incorrect.

Reply:
Yeah, I agree the helper has downsides too. But Roslyn could avoid using it when not targeting something with it available. For example, I imagine it could emit the equivalent of
```
ulong x = ...;
float f = x > long.MaxValue ? (float)(-x) + (float)0x8000000000000000 : (float)(long)x;
```
in those cases (or whatever the right way is to do the conversion manually). Signed long -> float conversion is representable with just `conv.r4`, I believe.

I don't think we need to try to fix this in .NET 9, but we may want to unify the debug/release behavior on something. We should probably open an issue for more discussion about the problem.

Reply:
> I don't think we need to try to fix this in .NET 9, but we may want to unify the debug/release behavior on something. We should probably open an issue for more discussion about the problem.

I've opened #106646 to track this. I think ARM64 was the only platform where Debug/Release behavior could diverge, due to const-folding always doing a two-step cast and `ucvtf/scvtf` being able to encode `ulong/long -> float` casts. Should we take this PR as-is for .NET 9?

Reply:
I expect this also repros on AVX512 capable hardware for x64.